### ISSUE TITLE
Android crashes after app is sent to background (running on device)

### ISSUE THREAD

**Issue by mmuil** [TheWidlarzGroup/react-native-video#1211]
### Current behavior

I'm using [https://github.com/aakashns/ReactMusic](https://github.com/aakashns/ReactMusic) to play remote audio files. When leaving the audio playing and exiting app, the audio plays for a few minutes (2-10) until it stops and is unresponsive to device's sound controls. 
When reopening  the app from recents, it reloads to app's initial state. 
Strangely, it also seems the issue appears only when the phone is disconnected from power source. When cable is attached, the audio can be played for hours. Some battery saving mechanism kicking in? For that reason there are no logcat findings. 

### Reproduction steps

Device: Pixel 2 8.1.0
react-native-video 3.0.0 3.1.0 3.2.0 all tested
react-native 0.55.4

1. Play audio
  `<Video source={{uri: track.audioUrl}}
  ref="audioElement"
  paused={this.state.paused}
  resizeMode="cover" 
  repeat={true}
  onLoadStart={this.loadStart} 
  onLoad={this.setDuration.bind(this)} 
  onProgress={this.setTime.bind(this)}
  onEnd={this.onEnd}
  onError={this.videoError}
  style={styles.audioElement} />`

2.  Put app to background
3. (Lock device)
4. Audio is stopped after a few minutes

### Expected behavior

Audio plays without interruptions until user interacts.

### Platform
Which player are you experiencing the problem on:

* Android ExoPlayer

No issues on iOS.

**Comment by mmuil** [comment]
Update:
Issue persists when react-native-video is the one and only installed component in the app. 
Code: 
```import React, {Component} from 'react';
import {StyleSheet, View,} from 'react-native';
import Video from 'react-native-video';

export default class App extends Component {
  render() {
    return (
      <View style={styles.container}>
          <Video source={require('./cats.mp3')}   // Can be a URL or a local file.
                 ref={(ref) => {
                     this.player = ref
                 }}
                 playInBackground={true}
                 playWhenInactive={true}
                 onBuffer={this.onBuffer}                // Callback when remote video is buffering
                 onEnd={this.onEnd}                      // Callback when playback finishes
                 onError={this.videoError}               // Callback when video cannot be loaded
          />
      </View>
    );
  }
}
```
Log pulled from device  (Pixel 2 8.1.0):

`ActivityManager: Killing 20894:com.videotest/u0a419 (adj 700): excessive cpu 53130 during 300068 dur=1762444 limit=2`

Seems like Android is pulling the app because of resources overuse? How can that be?
Issue still only appears when device is not attached to power outlet. With external power source, the crash does not happen.

**Comment by mmuil** [comment]
I see this has happened before: [#833](https://github.com/react-native-community/react-native-video/pull/833#issuecomment-339371245). Is background service for Exoplayer not implemented @cobarx ?

**Comment by mmjd** [comment]
any update on this?

**Comment by meiqi1992** [comment]
+1 how to resolve?

**Comment by meiqi1992** [comment]
@cobarx

**Comment by dkoo761** [comment]
+1

This seems like a real blocker (certainly for my use case of long-running audio meditations). I'm surprised there aren't more comments here. Wouldn't this basically block using react-native-video on pretty much ANY audio player app (podcasts, audio books, meditation, etc). If I can't find a workaround, I'll start looking into a fix myself this week and try to a create a PR if I can figure it out.

**Comment by dkoo761** [comment]
Forgot to mention, I'm running different device and versions:

- react-native: 0.57.8
- react-native-video: 4.2.0 - had to fallback to this version since later releases have blocker that prevents MP3 files from playing: https://github.com/react-native-community/react-native-video/issues/1442)
- device: Moto G5S Plus (Android 8.1.0)

Also, in my case the app suspends, but doesn't crash. If unlock my phone and go back to the app, it starts playing the audio exactly where it left off.

**Comment by dkoo761** [comment]
This issue can be reproduced in the Android Emulator without needing a physical device by going to "Extended Controls" in the emulator, clicking on Battery, then setting Charger Connection to None. That also makes logging easier than having to pull logs from your physical device.

**Comment by dkoo761** [comment]
I've managed to resolve this issue, at least for my use case, by following the ExoPlayer FAQ's guidelines for background audio: https://google.github.io/ExoPlayer/faqs.html#how-do-i-keep-audio-playing-when-my-app-is-backgrounded.

To implement wakelock and wifilock in RN, I used this package: "react-native-wakeful": "^0.1.0"

To implement a foreground service in RN, I used this package: "@voximplant/react-native-foreground-service": "^1.1.0"

Be sure to update AndroidManifest.xml as described in the READMEs for those 2 packages.

This demo repo's App.js has a clear example of how to use the foreground service: https://github.com/voximplant/react-native-foreground-service-demo/blob/master/App.js. The one thing missing from that example is that `stopService()` needs a guard at the start:
```
  async stopService() {
    // Only Android platform is supported
    if (Platform.OS !== 'android') {
      return;
    }

    await VIForegroundService.stopService();
  }
```

In my app, when the user clicks the Play/Pause button I do this:

```
    // play or pause the player, then.....

    if (this.isPaused) {
      // release wakelock, wifilock and stop foreground service
      wakeful.release();
      this.stopService();
    } else {
      // acquire wakelock, wifilock and start foreground service
      wakeful.acquire();
      this.startService();
    }
```

`wakeful.release()` / `wakeful.acquire()` can be left in for iOS as they simply no-op in that case.

If anyone needs any help, just let me know :)

**Comment by mmuil** [comment]
I've tested @dkoo761 's solution with my original example above and it indeed solves Android dismissing the app. 
However by this time [react-native-music-control](https://github.com/tanguyantoine/react-native-music-control) has been updated to create Android foreground service when it is enabled.
Basically, when `<Video>` is put in playing state, you should also put MusicControl (setNowPlaying) in playing state. MusicControl handles the foreground service and upon dismissing MusicControl, foreground service is also released. 
 It also has the benefit of having customisable media controls built in.

**Comment by flexgrip** [comment]
@dkoo761 

I tried your solution. I'm incredibly desperate to figure this out. It seems to be mostly a Samsung / Huawei issue. 

From what I can tell, those phones are just killing anything in the background consuming battery. I implemented your code above. However, I didn't notice anything. Like, should there be a notification on the screen or something whenever the foreground service starts? The app didn't do anything different, and it's still being killed in the background.

Where did you instantiate the `let wakeful = new Wakeful();`?

With your code, does it put a notification in the tray or the top of the screen when the app is in the background?

I'd pay money to find a damned solution to this. Android users are not thrilled. And they'll say, "Well YouTube doesn't have this problem." But I read on some android forums that samsung and huawei specifically target apps like YouTube and Whatsapp for whitelisting.
